src/backend/distributed/commands/schema.c

```diff
@@ -296,3 +323,78 @@ EnsureSequentialModeForSchemaDDL(void)
 							   "use only one connection for all future commands")));
 	SetLocalMultiShardModifyModeToSequential();
 }
+
+
+/*
+ * InvalidateFKeysForDistributedTablesInSchema takes a schema name and marks
+ * foreign graph to be invalidated, if a distributed table in the given schema
+ * has a foreign key connection.
+ */
+static void
+InvalidateFKeysForDistributedTablesInSchema(char *schemaName)
+{
+	ScanKeyData scanKey[1];
+	int scanKeyCount = 1;
+	Oid scanIndexId = InvalidOid;
+	bool useIndex = false;
+
+	Oid namespaceOid = get_namespace_oid(schemaName, true);
+
+	if (namespaceOid == InvalidOid)
+	{
+		return;
+	}
+
+	Relation pgClass = table_open(RelationRelationId, AccessShareLock);
+
+	ScanKeyInit(&scanKey[0], Anum_pg_class_relnamespace, BTEqualStrategyNumber,
+				F_OIDEQ, namespaceOid);
+	SysScanDesc scanDescriptor = systable_beginscan(pgClass, scanIndexId, useIndex, NULL,
+													scanKeyCount, scanKey);
+
+	HeapTuple heapTuple = systable_getnext(scanDescriptor);
+	while (HeapTupleIsValid(heapTuple))
+	{
+		Form_pg_class relationForm = (Form_pg_class) GETSTRUCT(heapTuple);
+		char *relationName = NameStr(relationForm->relname);
+		Oid relationId = get_relname_relid(relationName, namespaceOid);
+
+		/* we're not interested in non-valid, non-distributed relations */
+		if (relationId == InvalidOid || !IsCitusTable(relationId))
+		{
+			heapTuple = systable_getnext(scanDescriptor);
+			continue;
+		}
+
+		/* invalidate foreign key cache if the table involved in any foreign key */
+		if (TableReferenced(relationId) || TableReferencing(relationId))
+		{
+			MarkInvalidateForeignKeyGraph();
```

Comment:
Same as earlier, the original code had an early escape after the graph had been invalidated. Any reason we have changed that behaviour?